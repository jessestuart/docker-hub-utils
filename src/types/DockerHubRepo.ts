/**
 * This is a direct representation of what we get back from the `/repositories`
 * API call.
 */
export interface DockerHubAPIRepo {
  readonly can_edit: boolean
  readonly description: string
  readonly is_automated: boolean
  readonly is_migrated: boolean
  readonly is_private: boolean
  readonly last_updated: string
  readonly name: string
  readonly namespace: string
  readonly pull_count: number
  readonly repository_type: string
  readonly star_count: number
  readonly status: number
  readonly user: string
}

/**
 * Enum representing the architecture defined in part of an OCI image's
 * manifest list.
 *
 * There are, of course, a shit ton of other supported architectures.
 * These are just the ones I've found to be most relevant / common.
 */
export enum Architecture {
  amd64 = 'amd64',
  arm = 'arm',
  arm64 = 'arm64',
}

export enum ManifestMediaType {
  Manifest = 'application/vnd.docker.distribution.manifest.v2+json',
  ManifestList = 'application/vnd.docker.distribution.manifest.list.v2+json',
}

/**
 * Yes, there's *way* more information contained in the manifest / "fat"
 * manifestList than just architectures, but I find this to be the most
 * relevant section for my projects. PR's welcome.
 */
export interface DockerManifest {
  readonly digest: string
  readonly mediaType: ManifestMediaType
  readonly platform: Array<{
    architecture: Architecture
  }>
  readonly schemaVersion: 1 | 2 | any
}

export interface DockerHubRepo {
  // ========================
  // Main fields of interest
  // ========================
  readonly description: string | null | undefined
  readonly lastUpdated: Date
  readonly name: string
  readonly pullCount: number
  readonly starCount: number
  readonly user: string

  manifest?: DockerManifest

  // =============================================
  // Other stuff that comes down through the API,
  // that some may find useful
  // =============================================
  readonly canEdit: boolean
  readonly isAutomated: boolean
  readonly isMigrated: boolean
  readonly isPrivate: boolean
  readonly namespace: string
  readonly repositoryType: string
  readonly status: number
}
