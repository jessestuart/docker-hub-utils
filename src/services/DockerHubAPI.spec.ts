import axios from 'axios'
import { DateTime } from 'luxon'
import R from 'ramda'

import manifestFixture from '../__tests__/manifest_fixture.json'
import fixture from '../__tests__/repos_fixture.json'
import repoFixturesInvalid from '../__tests__/repos_fixtures_invalid.json'
import { DockerHubRepo } from '../types/DockerHubRepo'
import log from '../utils/log'
import {
  extractRepositoryDetails,
  fetchManifestList,
  queryTopRepos,
} from './DockerHubAPI'

const get = jest.spyOn(axios, 'get')
const logInfo = jest.spyOn(log, 'info')

const repoFixtures = fixture.data.results

describe('DockerHub handler', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })

  test('extractRepositoryDetails for a non-existant repo', () => {
    const repositoryDetails = extractRepositoryDetails([])
    expect(repositoryDetails).toHaveLength(0)
  })

  test('extractRepositoryDetails happy path', () => {
    const repoResults = repoFixtures
    const repositoryDetails = extractRepositoryDetails(repoResults)
    expect(repositoryDetails).toHaveLength(R.length(repoResults))
    const keys = [
      'description',
      'lastUpdated',
      'name',
      'pullCount',
      'starCount',
    ]

    // Verify values are defined for each of the above keys.
    repositoryDetails.forEach(repo => {
      keys.forEach((key: string) => {
        expect(repo).toHaveProperty(key)
      })
    })
  })

  test('queryTopRepos', async () => {
    // Configure axios to return our fixtures, rather than hitting the API.
    get.mockResolvedValueOnce(fixture)

    const topRepos: DockerHubRepo[] = await queryTopRepos({
      numRepos: 25,
      user: 'jessestuart',
    })

    // Only one HTTP call required -- no auth needed when querying repos.
    expect(get).toHaveBeenCalledTimes(1)
    expect(topRepos).toHaveLength(R.length(repoFixtures))

    const totalPulls = R.sum(R.pluck('pullCount', topRepos))
    expect(totalPulls).toMatchSnapshot()

    const topRepoName = R.pipe(
      R.head,
      R.path(['name']),
    )(topRepos)

    expect(topRepoName).toBe('owntracks')
  })

  test('queryTopRepos, limit results by date', async () => {
    get.mockResolvedValueOnce(fixture)

    const topRepos: DockerHubRepo[] = await queryTopRepos({
      lastUpdatedSince: DateTime.fromISO('2019-06-05'),
      numRepos: 25,
      user: 'jessestuart',
    })

    expect(get).toHaveBeenCalledTimes(1)
    expect(topRepos).toMatchSnapshot()
    expect(topRepos.length).toBeLessThan(repoFixtures.length)
  })

  test('queryTopRepos with invalid data (lastUpdated undefined)', async () => {
    get.mockResolvedValueOnce(repoFixturesInvalid)

    const topRepos: DockerHubRepo[] = await queryTopRepos({
      lastUpdatedSince: DateTime.fromISO('2019-06-05'),
      numRepos: 25,
      user: 'jessestuart',
    })

    expect(get).toHaveBeenCalledTimes(1)
    expect(topRepos).toMatchSnapshot()
    expect(topRepos.length).toBe(0)
  })

  test('queryTopRepos with no numRepos specified', async () => {
    get.mockResolvedValueOnce(fixture)

    const topRepos: DockerHubRepo[] = await queryTopRepos({
      lastUpdatedSince: DateTime.fromISO('2019-06-05'),
      user: 'jessestuart',
    })

    expect(get).toHaveBeenCalledTimes(1)
    expect(topRepos).toMatchSnapshot()
  })

  test('queryTopRepos fails if >100 repos requested.', () => {
    get.mockResolvedValueOnce(fixture)

    const topReposRequest = queryTopRepos({
      numRepos: 101,
      user: 'jessestuart',
    })
    expect(topReposRequest).rejects.toThrow()

    expect(get).toHaveBeenCalledTimes(0)
  })

  test('fetchManifestList happy path.', async () => {
    get.mockResolvedValueOnce({ data: { token: 'FAKE_TOKEN' } })
    get.mockResolvedValueOnce(manifestFixture)

    const topRepo: DockerHubRepo = R.head(
      extractRepositoryDetails(repoFixtures),
    )
    if (!topRepo) {
      fail('Unable to load repos fixture.')
      return
    }

    const manifestList = await fetchManifestList(topRepo)
    expect(get).toHaveBeenCalledTimes(2)
    expect(get.mock.calls).toMatchSnapshot()
    expect(get).toHaveBeenNthCalledWith(1, 'https://auth.docker.io/token', {
      params: {
        scope: 'repository:jessestuart/owntracks:pull',
        service: 'registry.docker.io',
      },
    })
    expect(get).toHaveBeenNthCalledWith(
      2,
      'https://registry-1.docker.io/v2/jessestuart/owntracks/manifests/latest',
      {
        headers: {
          Accept: 'application/vnd.docker.distribution.manifest.list.v2+json',
          Authorization: 'Bearer FAKE_TOKEN',
        },
      },
    )

    expect(manifestList).toMatchSnapshot()
  })

  // eslint-disable-next-line
  test('fetchManifestList returns early for schemaVersion 1.', async () => {
    get.mockResolvedValueOnce({ data: { token: 'FAKE_TOKEN' } })
    get.mockResolvedValueOnce({ data: { schemaVersion: 1 } })

    const topRepo: DockerHubRepo = R.head(
      extractRepositoryDetails(repoFixtures),
    )

    const manifestList = await fetchManifestList(topRepo)
    expect(manifestList).toBeUndefined()
    expect(logInfo).toHaveBeenCalledTimes(1)
  })

  // eslint-disable-next-line
  test('fetchManifestList fails when no token returned.', async () => {
    get.mockResolvedValueOnce({ data: {} })
    get.mockResolvedValueOnce(manifestFixture)

    const topRepo = R.head(extractRepositoryDetails(repoFixtures))
    if (!topRepo) {
      fail('Unable to load repos fixture.')
      return
    }

    expect(fetchManifestList(topRepo)).rejects.toThrow()
    expect(get).toHaveBeenCalledTimes(1)
  })
})
