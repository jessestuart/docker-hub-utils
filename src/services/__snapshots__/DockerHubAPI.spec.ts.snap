// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerHub handler fetchManifestList happy path. 1`] = `
Array [
  Array [
    "https://auth.docker.io/token",
    Object {
      "params": Object {
        "scope": "repository:jessestuart/owntracks:pull",
        "service": "registry.docker.io",
      },
    },
  ],
  Array [
    "https://registry-1.docker.io/v2/jessestuart/owntracks/manifests/latest",
    Object {
      "headers": Object {
        "Accept": "application/vnd.docker.distribution.manifest.list.v2+json",
        "Authorization": "Bearer FAKE_TOKEN",
      },
    },
  ],
]
`;

exports[`DockerHub handler fetchManifestList happy path. 2`] = `
Object {
  "manifests": Array [
    Object {
      "digest": "sha256:d0593f2bf35b854611dda94b5210243f8473847150d1709b12163453841bf5f8",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "amd64",
        "os": "linux",
      },
      "size": 1365,
    },
    Object {
      "digest": "sha256:0ec4c38328ab0e0b8798c28384fd44bab385b8313b13aa8f494e940e83a12d69",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "arm",
        "os": "linux",
      },
      "size": 1369,
    },
    Object {
      "digest": "sha256:72d730063d55fa164378188f92119e00769dbc759f95cbf1319705e90d6214ea",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "arm64",
        "os": "linux",
      },
      "size": 1369,
    },
  ],
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "schemaVersion": 2,
}
`;

exports[`DockerHub handler queryTopRepos 1`] = `1718496`;

exports[`DockerHub handler queryTopRepos with invalid data (lastUpdated undefined) 1`] = `Array []`;

exports[`DockerHub handler queryTopRepos with no numRepos specified 1`] = `
Array [
  Object {
    "canEdit": false,
    "description": "OwnTracks. Multiarch. ðŸ‘Œ",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:53.188Z",
    "name": "owntracks",
    "namespace": "jessestuart",
    "pullCount": 488928,
    "repositoryType": "image",
    "starCount": 3,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "The every(wo)man's reverse proxy, on all your boxes or boards. amd64, arm64, & armhf nightly builds.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-17T06:13:12.702Z",
    "name": "caddy",
    "namespace": "jessestuart",
    "pullCount": 363038,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Nightly multi-architecture (amd64, arm64, armv7) builds for the Kubernetes Helm tiller service.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:23.660Z",
    "name": "tiller",
    "namespace": "jessestuart",
    "pullCount": 335171,
    "repositoryType": "image",
    "starCount": 10,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "All the dependencies, up-to-date, all the time. Nightly multiarch builds. (amd64 + arm64)",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:00.434Z",
    "name": "renovate",
    "namespace": "jessestuart",
    "pullCount": 143411,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Minio server â€” supports arm (arm32v6, arm32v7, armhf) arm64 (aarch64), and amd64 architectures.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:28.724Z",
    "name": "minio",
    "namespace": "jessestuart",
    "pullCount": 74836,
    "repositoryType": "image",
    "starCount": 3,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "An OSS metrics dashboard and graph editor for Graphite, Elasticsearch, Prometheus and InfluxDB.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:46.783Z",
    "name": "grafana",
    "namespace": "jessestuart",
    "pullCount": 52749,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "amd64, arm64, and armhf Docker images for Minio Client.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:01.207Z",
    "name": "mc",
    "namespace": "jessestuart",
    "pullCount": 45199,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Nightly multiarch (arm64, armhf, amd64) builds of Heptio Velero for Kubernetes.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:27.980Z",
    "name": "velero",
    "namespace": "jessestuart",
    "pullCount": 36522,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "kube2iam provides different AWS IAM roles for pods running on Kubernetes. 
Nightly multiarch builds.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:11.724Z",
    "name": "kube2iam",
    "namespace": "jessestuart",
    "pullCount": 11239,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:16:41.077Z",
    "name": "elasticsearch_exporter",
    "namespace": "jessestuart",
    "pullCount": 10668,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:38.629Z",
    "name": "docker-hub-exporter",
    "namespace": "jessestuart",
    "pullCount": 8596,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:15:25.027Z",
    "name": "homemaker",
    "namespace": "jessestuart",
    "pullCount": 4527,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:38.734Z",
    "name": "pihole_exporter",
    "namespace": "jessestuart",
    "pullCount": 3271,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:03:30.885Z",
    "name": "vagrant-build-base",
    "namespace": "jessestuart",
    "pullCount": 2509,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
]
`;

exports[`DockerHub handler queryTopRepos, limit results by date 1`] = `
Array [
  Object {
    "canEdit": false,
    "description": "OwnTracks. Multiarch. ðŸ‘Œ",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:53.188Z",
    "name": "owntracks",
    "namespace": "jessestuart",
    "pullCount": 488928,
    "repositoryType": "image",
    "starCount": 3,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "The every(wo)man's reverse proxy, on all your boxes or boards. amd64, arm64, & armhf nightly builds.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-17T06:13:12.702Z",
    "name": "caddy",
    "namespace": "jessestuart",
    "pullCount": 363038,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Nightly multi-architecture (amd64, arm64, armv7) builds for the Kubernetes Helm tiller service.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:23.660Z",
    "name": "tiller",
    "namespace": "jessestuart",
    "pullCount": 335171,
    "repositoryType": "image",
    "starCount": 10,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "All the dependencies, up-to-date, all the time. Nightly multiarch builds. (amd64 + arm64)",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:00.434Z",
    "name": "renovate",
    "namespace": "jessestuart",
    "pullCount": 143411,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Minio server â€” supports arm (arm32v6, arm32v7, armhf) arm64 (aarch64), and amd64 architectures.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:28.724Z",
    "name": "minio",
    "namespace": "jessestuart",
    "pullCount": 74836,
    "repositoryType": "image",
    "starCount": 3,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "An OSS metrics dashboard and graph editor for Graphite, Elasticsearch, Prometheus and InfluxDB.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:46.783Z",
    "name": "grafana",
    "namespace": "jessestuart",
    "pullCount": 52749,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "amd64, arm64, and armhf Docker images for Minio Client.",
    "isAutomated": true,
    "isMigrated": true,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:01.207Z",
    "name": "mc",
    "namespace": "jessestuart",
    "pullCount": 45199,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "Nightly multiarch (arm64, armhf, amd64) builds of Heptio Velero for Kubernetes.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:27.980Z",
    "name": "velero",
    "namespace": "jessestuart",
    "pullCount": 36522,
    "repositoryType": "image",
    "starCount": 1,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "kube2iam provides different AWS IAM roles for pods running on Kubernetes. 
Nightly multiarch builds.",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:11.724Z",
    "name": "kube2iam",
    "namespace": "jessestuart",
    "pullCount": 11239,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:16:41.077Z",
    "name": "elasticsearch_exporter",
    "namespace": "jessestuart",
    "pullCount": 10668,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:26:38.629Z",
    "name": "docker-hub-exporter",
    "namespace": "jessestuart",
    "pullCount": 8596,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:15:25.027Z",
    "name": "homemaker",
    "namespace": "jessestuart",
    "pullCount": 4527,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:25:38.734Z",
    "name": "pihole_exporter",
    "namespace": "jessestuart",
    "pullCount": 3271,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
  Object {
    "canEdit": false,
    "description": "",
    "isAutomated": false,
    "isMigrated": false,
    "isPrivate": false,
    "lastUpdated": "2019-07-26T00:03:30.885Z",
    "name": "vagrant-build-base",
    "namespace": "jessestuart",
    "pullCount": 2509,
    "repositoryType": "image",
    "starCount": 0,
    "status": 1,
    "user": "jessestuart",
  },
]
`;
